psl-core/src/main/java/org/linqs/psl/util/ModelDataCollector.java

```diff
@@ -0,0 +1,332 @@
+    }
+
+    // Adds a ground rules dissatisfaction / infeasibility.
+    public static void dissatisfactionPerGroundRule(GroundRuleStore groundRuleStore) {
+        for (GroundRule groundRule : groundRuleStore.getGroundRules()) {
+            String strGroundRuleId = Integer.toString(System.identityHashCode(groundRule));
+            Map<String, Object> groundRuleObj = modelData.groundRules.get(strGroundRuleId);
+            if (groundRule instanceof WeightedGroundRule) {
+                WeightedGroundRule weightedGroundRule = (WeightedGroundRule) groundRule;
+                groundRuleObj.put("dissatisfaction", weightedGroundRule.getIncompatibility());
+            } else {
+                UnweightedGroundRule unweightedGroundRule = (UnweightedGroundRule) groundRule;
+                groundRuleObj.put("dissatisfaction", unweightedGroundRule.getInfeasibility());
+            }
+        }
+    }
+
+    // Decorates an entry in a Formula to have constants or negations marked.
+    public static Object decorateFormula(String groundAtom, GroundRule groundRule, boolean negation){
+        for (GroundAtom atom : groundRule.getAtoms()) {
+            if (groundAtom.contains(atom.toString())){
+                if (negation) {
+                    Integer[] groundAtomObj = {System.identityHashCode(atom), 1};
+                    return groundAtomObj;
+                } else {
+                    return System.identityHashCode(atom);
+                }
+            }
+        }
+        return null;
+    }
+
+    // Parses through an atom object, and fills out variables with their proper values.
+    public static String parseAtom (Formula f, Map<String, String> varConstMap) {
+        Atom atom = (Atom) f;
+        String groundedAtom = atom.toString();
+        Term[] arguments = atom.getArguments();
+        for (Term t : arguments){
+            if (t instanceof Variable) {
+                // Only want to replace terms, to do so make sure they are in a parenthesis.
+                String leftVarParen = "(" + t.toString();
+                String rightVarParen = t.toString() + ")";
+                if (groundedAtom.contains(leftVarParen)) {
+                    String replacement = "(\'" + varConstMap.get(t.toString()) + "\'";
+                    groundedAtom = groundedAtom.replace(leftVarParen, replacement);
+                } else if (groundedAtom.contains(rightVarParen)) {
+                    String replacement = "\'" + varConstMap.get(t.toString()) + "\')";
+                    groundedAtom = groundedAtom.replace(rightVarParen, replacement);
+                }
+            }
+        }
+        return groundedAtom;
+    }
+
+    // Parses through a formula object, and creates the needed object for modelData object consumption.
+    public static ArrayList<Object> parseFormula(Formula f, Map<String, String> varConstMap, GroundRule groundRule, boolean negation) {
+        ArrayList<Object> groundAtoms = new ArrayList<Object>();
```

Comment:
Unless you specifically need an ArrayList, use a List.
In general, you want to keep your types as general as possible so people have flexibility.
Here we are just using normal list functionality, and don't actually care if we are using an ArrayList, LinkedList, or some other type of list.